va/http.go

```diff
@@ -0,0 +1,424 @@
+		Hostname:          target.host,
+		Port:              strconv.Itoa(target.port),
+		AddressesResolved: target.available,
+	}
+
+	// Build a URL with the target's IP address and port
+	targetIP, err := target.IP()
+	if err != nil {
+		return nil, record, err
+	}
+	record.AddressUsed = *targetIP
+	url := httpValidationURL(*targetIP, target.path, target.port)
+	record.URL = url.String()
+
+	// If there's no provided HTTP request to mutate (e.g. a redirect request
+	// we're following as part of a validation) then construct a new initial HTTP
+	// GET request for the validation.
+	if req == nil {
+		req, err = http.NewRequest("GET", url.String(), nil)
+		if err != nil {
+			return nil, record, err
+		}
+		// Immediately reconstruct the request using the validation context
+		req = req.WithContext(ctx)
+		if va.userAgent != "" {
+			req.Header.Set("User-Agent", va.userAgent)
+		}
+		// Some of our users use mod_security. Mod_security sees a lack of Accept
+		// headers as bot behavior and rejects requests. While this is a bug in
+		// mod_security's rules (given that the HTTP specs disagree with that
+		// requirement), we add the Accept header now in order to fix our
+		// mod_security users' mysterious breakages. See
+		// <https://github.com/SpiderLabs/owasp-modsecurity-crs/issues/265> and
+		// <https://github.com/letsencrypt/boulder/issues/1019>. This was done
+		// because it's a one-line fix with no downside. We're not likely to want to
+		// do many more things to satisfy misunderstandings around HTTP.
+		req.Header.Set("Accept", "*/*")
+	}
+
+	// Override the request's target URL and Host
+	req.URL = url
+	req.Host = target.host
+	return req, record, nil
+}
+
+// fetchHTTPSimple invokes processHTTPValidation and if an error result is
+// returned, converts it to a problem. Otherwise the results from
+// processHTTPValidation are returned.
+func (va *ValidationAuthorityImpl) fetchHTTPSimple(
+	ctx context.Context,
+	host string,
+	path string) ([]byte, []core.ValidationRecord, *probs.ProblemDetails) {
+	body, records, err := va.processHTTPValidation(ctx, host, path)
+	if err != nil {
+		// Use detailedError to convert the error into a problem
+		return body, records, detailedError(err)
+	}
+	return body, records, nil
+}
+
+// processHTTPValidation performs an HTTP validation for the given host, port
+// and path. If successful the body of the HTTP response is returned along with
+// the validation records created during the validation. If not successful
+// a non-nil error and potentially some ValidationRecords are returned.
+func (va *ValidationAuthorityImpl) processHTTPValidation(
+	ctx context.Context,
+	host string,
+	path string) ([]byte, []core.ValidationRecord, error) {
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		// Shouldn't happen: All requests should have a deadline by this point.
+		deadline = time.Now().Add(100 * time.Second)
+	} else {
+		// Set the context deadline slightly shorter than the HTTP deadline, so we
```

Comment:
This code came from our custom dialer, and made the assumption that its context would only apply to dialing. In this PR, the adjusted context winds up getting used for both the dial and the HTTP fetch, so this comment doesn't make sense anymore. Also, I think the switch means we may have lost the more-specific errors.

For context, in https://github.com/letsencrypt/boulder/pull/3629 we made it so you can get either a "timeout during connect" or a "timeout after connect."

I think we can achieve the same effect in the new code by defining a `DialContext` on our custom `Transport`. However, instead of the complex logic we had before, this `DialContext` would merely shave 10ms off the context it was given before passing it along to the real `DialContext` (or in other words, this block of code would move to our custom `DialContext` implementation).

Reply:
That sounds reasonable :+1: I wasn't 100% sure it made sense to port this "shave" logic into the fetch.